extensions/csrc/cuda/utils/vec_type_traits.h

```diff
@@ -13,70 +13,27 @@ namespace utils {
 template <typename T, int VecSize>
 struct VecTypeTrait {};
 
-template <typename T>
-struct VecTypeTrait<T, 1> {
-  using Type = T;
-};
-
-template <>
-struct VecTypeTrait<c10::BFloat16, 2> {
-  using Type = float;
-};
-
-template <>
-struct VecTypeTrait<c10::BFloat16, 4> {
-  using Type = float2;
-};
-
-template <>
-struct VecTypeTrait<c10::BFloat16, 8> {
-  using Type = float4;
-};
-
-template <>
-struct VecTypeTrait<c10::Half, 2> {
-  using Type = float;
-};
-
-template <>
-struct VecTypeTrait<c10::Half, 4> {
-  using Type = float2;
-};
-
-template <>
-struct VecTypeTrait<c10::Half, 8> {
-  using Type = float4;
-};
-
-template <>
-struct VecTypeTrait<float, 2> {
-  using Type = float2;
-};
-
-template <>
-struct VecTypeTrait<float, 4> {
-  using Type = float4;
-};
-
-template <>
-struct VecTypeTrait<float, 8> {
-  using Type = float4;
-};
-
-template <>
-struct VecTypeTrait<uint8_t, 2> {
-  using Type = half;
-};
-
-template <>
-struct VecTypeTrait<uint8_t, 4> {
-  using Type = half2;
-};
-
-template <>
-struct VecTypeTrait<uint8_t, 8> {
-  using Type = float2;
-};
+#define VEC_TYPE_TRAITS_SPECIALIZATION(T, VEC_SIZE, VECT, ARGS...) \
+  template <ARGS>                                                  \
+  struct VecTypeTrait<T, VEC_SIZE> {                               \
+    using Type = VECT;                                             \
+  };
+
+VEC_TYPE_TRAITS_SPECIALIZATION(T, 1, T, typename T)
+VEC_TYPE_TRAITS_SPECIALIZATION(c10::BFloat16, 2, float)
+VEC_TYPE_TRAITS_SPECIALIZATION(c10::BFloat16, 4, float2)
+VEC_TYPE_TRAITS_SPECIALIZATION(c10::BFloat16, 8, float4)
+VEC_TYPE_TRAITS_SPECIALIZATION(c10::Half, 2, float)
+VEC_TYPE_TRAITS_SPECIALIZATION(c10::Half, 4, float2)
+VEC_TYPE_TRAITS_SPECIALIZATION(c10::Half, 8, float4)
+VEC_TYPE_TRAITS_SPECIALIZATION(float, 2, float2)
+VEC_TYPE_TRAITS_SPECIALIZATION(float, 4, float4)
+VEC_TYPE_TRAITS_SPECIALIZATION(float, 8, float4)
```

Comment:
Will this case be actually be used? It seems to have different way of usage compared with other mappings.

Reply:
`VEC_TYPE_TRAITS_SPECIALIZATION(float, 8, float4)`

Reply:
I think the same as you, here is the usage
![image](https://github.com/hpcaitech/ColossalAI/assets/28504079/022cd8b8-99b9-4879-856a-5c4adf64da5d)
![image](https://github.com/hpcaitech/ColossalAI/assets/28504079/5425aa82-1f65-4363-a7a5-b7f993d5cf67)
Since you brought it up, I think this should be deleted